// MyModules/folderCreator/index.js
var fs=require('fs');

function createFolder(path,callback){
    if(!fs.existsSync(path)){
        fs.mkdirSync(path, 0766, function(err){

            if(err){
                //console.log(err);
                callback(false);
            }
            callback(true);
        });

    }
}
function haveFolder(path,callback){
    return fs.existsSync(path);
}
function createFile(path,text,callback){
    fs.writeFile(path, text, function(err) {
        if(err) {
            if(callback)callback(false);
            return;
        }
        if(callback)callback(true);
    });
}
function writeFile(path,text,callback){
    fs.writeFile(path, text, function (err) {
        if (err) return console.log(err);
     //   console.log('Hello World > helloworld.txt');
    });
}
function read(path,callback) {
    fs.readFile(path,callback );
}
function addToF(path,text,callback){
    read(path,function(er,data){
       // console.log(data.toString());
        writeFile(path,data.toString()+'\n'+text,callback);
    })
}

module.exports.add=addToF;
module.exports.read=read;
module.exports.wrFile=writeFile;
module.exports.crFile=createFile;
module.exports.hvFolder=haveFolder;
module.exports.crFolder=createFolder;
//console.log(__dirname);
