// ADS/SKR/index.js
/**
 * Created by serne_000 on 17.12.2015.
 */
var name;
var deb=require('DEBUG').debug;

function debug(text){
    deb(text,name);
}
exports.getN=function(n){
    name=n;
};
var reader;
var sendMess;
var sendStick;
exports.getStick=function(s){
    sendStick=s;
};
exports.getMess=function(m){
   // console.log(m);
    sendMess=m;
};
exports.getR=function(r){
    reader=r;
};
exports.mainF=function(text,id){
   // console.log('here');
   // console.log(sendMess);
    sendMess(id, "[Hook: PHARAON] х2 \n" +
        " Skrrt х4 в мёртвых nike'ах \n" +
        " Skrrt х4 в белой майке \n" +
        " Skrrt х4 чёрный сталкер \n" +
        " Skrrt х4 с зипом сканка \n" +
        " [Verse 1: PHARAON] \n" +
        "  Мой узи в снегу, \n" +
        "он сплюнет их в могилы (пау х4) \n" +
        " Тварь, имей в виду, \n" +
        " я - в камуфляжном ЗИЛ-е (катаюсь) \n" +
        " Я считаю бирки, на мне чёрный трэшер\n " +
        "Томми Хилфигер, Волга, smoking session \n" +
        "Купола на солнце, моментально скручен \n" +
        "Десять стильных сучек, но мне с ними скучно\n " +
        "Все друзья ублюдки, мы всегда за кадром\n" +
        " Я ушёл от марок, не вернусь обратно\n" +
        " Три зелёных стрелки вокруг её шеи (Dead Dynasty)\n" +
        " Она хочет мёртвых, сука, неужели?\n" +
        " В ледяной гробнице в белых платьях нимфы\n" +
        " Засыпал в могиле со стволом в обнимку\n" +
        " Я всегда расстроен, но мой blunt неистов\n" +
        " Моя crew прекрасна, как премьер-министр\n" +
        " На запястье шрамы (эмо), заряжаю Магнум\n" +
        " Dynasty сияет, я взрываю пальмы\n"+
        " Skrrt х4 в белой майке \n" +
        " Skrrt х4 чёрный сталкер \n" +
        " Skrrt х4 с зипом сканка\n" );
};
exports.info=function(){
    return 'СКЁЁЁЁЁЁЁЁР!';
};
module.exports.addF=function(){
    return {};
};
var sendPhoto;
exports.getPhoto=function(m){
    sendPhoto=m;
};
