// SERVER/requestHandlers.js
var reader = require("./reader");
var createObj=require("DB/createObj.js");
var logPass=require("./validLogPass.js");
var online=require("DB/onlineUsers.js");

//var config=require("../config/config.js");

function login(res,params){
    reader.sendFileSave('./login.html',res);

};
function register(res,params){
        reader.sendFileSave('./register.html', res);
};
function reg(res,params){
  //  console.log(params);
    var a=logPass.login(params["login"]);
    if(a){
        res.end(a);
        return;
    }
    a=logPass.pass(params["password"]);
    if(a){
        res.end(a);
        return;
    }
    createObj.createUser(params["login"],params["password"],function(err){
        if(err){
            console.log(err);
            res.end("Такой пользователь уже зарегистрирован");
            return;
        }
        res.end("Успешно зарегистрирован!");
    });
    //res.end('Успешно зарегистрирован');
};
function log(res,params){
    var user;
    createObj.getUsers({'username':params['login']},function(users){
        if(users.length){// console.log("EEEE")
            user = users[0];
            if(user.checkPassword(params['password'])){
                res.end("");
                if(!online.find(params['login']))
                online.onlineUser(user.username);
            }
            else
            res.end("Неправильный логин или пароль");
        }
        else{
            res.end("Неправильный логин или пароль");
        }
    });

  //  res.end("ЩАс доделаю");
};
function main(res,params){
    reader.sendFileSave('./main.html', res);
};
function exit(res,params){


}

//Chat



exports.main=main;
exports.log=log;
exports.reg=reg;
exports.register=register;
exports.login=login;
