// ADS/BALDA/index.js
//ROOT
var name;
var deb=require('DEBUG').debug;

function debug(text){
    deb(text,name);
}
exports.getN=function(n){
    name=n;
};
{var reader;
var sendMess;
var sendStick;
exports.getStick=function(s){
    sendStick=s;
};
exports.getMess=function(m){
    sendMess=m;
};
var sendPhoto;
exports.getPhoto=function(m){
    sendPhoto=m;
};
exports.getR=function(r){
    reader=r;
};
exports.mainF=mainF;
exports.info=function(){

}
module.exports.addF=function(){
    return {};
};
var sendPhoto;
exports.getPhoto=function(m){
    sendPhoto=m;
};}
var words=require('./nouns.json');
var games={};
function game(id){
    //this.word=word;
    var word=randomWord();
    this.id=id;
    this.maxMistakes=5;
    this.mistakes=0;
    var a=[];
    for(var i=0;i<word.length;i++){
        a[i]=new letter(word[i]);
    }
    this.word=a;
    this.usedLets=[];
   // console.log(a);
}
function letter(let,op){
    this.let=let;
    this.open=op||false;
}
function read(id) {
    var res = {};
    sendMess(id, 'Введите букву');
    res.f = function (text) {
        var r = {};
        r.del = true;
        mainF(text, id);
        return r;
        //console.log(text);
    };
    return res;
}
function mainF(text,id){
    text=text.toLowerCase();
    if(games[id]){
        var g=games[id];
        if (text && text.length===1) {
            text=text.replace(/ /g,'');
            sendMess(id,g.turn(text));
        } else {
            reader(read(id),id);
        }
    }else {
        var g=new game(id);
        games[id]=g;
        sendMess(id,'Игра создана');
    }
}
//console.log(words);
game.prototype.turn=function(let){
    var needC=[];
    for(var i=0;i<this.word.length;i++){
        if(this.word[i].let===let){
            needC.push(i);
        }
    }
    //console.log(needC);
    if(needC.length){
        for(var i in needC){
            this.word[needC[i]].open=true;
        }


        //var r=false;
        for(var i=0;i<this.word.length;i++){
            if(!this.word[i].open){
                return this.log();
            }
        }
        var res='Вы угадали слово! \n';
        for(var i=0;i<this.word.length;i++){
            res+=this.word[i].let;
        }
        delete games[this.id];
        return res;

    }else{

        this.mistakes++;
        if(this.mistakes>=this.maxMistakes){
            var res='Вы проиграли! \n Это было слово ';
            for(var i=0;i<this.word.length;i++){
                res+=this.word[i].let;
            }
            delete games[this.id];
            return res;
        }else{
            return 'Вы не угадали! \n' +this.log();
        }
    }
};
game.prototype.log=function(){
    var res=this.mistakes+'x ошибок \n';

    for(var i in this.word){
        this.word[i].open?res+=this.word[i].let+' ':res+='? ';
    }
    return res;
};
function randomWord(){
    var lengths=[];
    for(var i in words) {
        if(i*1>6)
        lengths.push(i);
    }
    var l=Math.floor(Math.random()*lengths.length);
    var n=Math.floor(Math.random()*words[lengths[l]].length);
    return words[lengths[l]][n];
}
//console.log(randomWord());
//var g=new game('залупа',1488);
//g.turn('z');
//console.log(g.turn('м'));
//console.log(g.turn('м'));
//console.log(g.turn('м'));
//console.log(g.turn('м'));
//console.log(g.turn('м'));
//console.log(g.log());
