// ADS/ANSWER/index.js
/**
 * Created by serne_000 on 18.12.2015.
 */
var name;
var deb=require('DEBUG').debug;

function debug(text){
    deb(text,name);
}
exports.getN=function(n){
    name=n;
};
var reader;

var phrases=[];
function addPhrase(phr){
    phrases.push(phr);
};
addPhrase('Конечно');
addPhrase('Врятли');
addPhrase('Не уверен');
addPhrase('Kappa');
addPhrase('Точно нет!');
addPhrase('Не знаю');
addPhrase('КееееееК');

exports.getR=function(r){
    reader=r;
};
var sendMess;
var sendStick;
exports.getStick=function(s){
    sendStick=s;
};
exports.getMess=function(m){
    sendMess=m;
};
function read(id){
    var res={};
        sendMess(id, 'На какой вопрос ответить?');
        res.f=function(text){
        var r={};
        r.del=true;
        mainF(text,id);
        return r;
    };
    return res;
}
function mainF(text,id){
    // console.log('here');
    // console.log(phrases);
    if(text)
        sendMess(id,phrases[Math.floor(Math.random()*phrases.length)]);
    else {
        var res= read(id);
        reader(res,id);
    }
};
exports.mainF=mainF;
exports.info=function(){
    return 'Отвечает на ваши копросы';
};
module.exports.addF=function(){
    return {};
};
var sendPhoto;
exports.getPhoto=function(m){
    sendPhoto=m;
};
