// ADS/WIKIPEDIA/index.js
var name;
var deb=require('DEBUG').debug;

function debug(text){
    deb(text,name);
}
exports.getN=function(n){
    name=n;
};
var sendMess;
var sendStick;
var reader;
exports.getStick = function (s) {
    sendStick = s;
};
exports.getMess = function (m) {
    sendMess = m;
};
var http = require('http');
var url = require('url');
var request = require('request');
var cheerio = require('cheerio');
var urlencode = require('urlencode');
function getById(body, id) {
    var $ = cheerio.load(body);
    return $('#' + id);
}
function getUrl(name) {
    name = urlencode(name);
    return "https://ru.wikipedia.org/w/index.php?search=" + name + "&title=%D0%A1%D0%BB%D1%83%D0%B6%D0%B5%D0%B1%D0%BD%D0%B0%D1%8F%3A%D0%9F%D0%BE%D0%B8%D1%81%D0%BA&go=%D0%9F%D0%B5%D1%80%D0%B5%D0%B9%D1%82%D0%B8";

}
function read1(id,callback,ar) {
    var res = {};
    res.f = function (text) {
       // console.log(text);
        var r = {};

        r.del = !!ar[text-1];

        if(r.del && (Number.isInteger(text * 1)) ){
            console.log('!');
            find(ar[text*1-1]['name'],id,callback,ar[text*1-1]['url']);
        } else{
            sendMess(id,'Неправильный ввод!');
        }
        return r;
    };
    return res;
}

function selection(body, id,callback) {
    var $ = cheerio.load(body);
    var a = $('.searchresults').find($('.mw-search-results')).find("li");
    var fst;
    var res = [];
    for (var i in a) {
        if (!Number.isInteger(i * 1))continue;
        res.push({'name': '', 'url': ''});
        fst = a[i];
        fst = $(fst).find($('.mw-search-result-heading')).find('a');

        res[res.length - 1]['name'] =$(fst).attr('title');
        res[res.length - 1]['url'] ='https://ru.wikipedia.org'+$(fst).attr('href') /*getUrl(res[res.length - 1]['name'])*/;

    }
    var str = 'Нужной статьи не нашлось, но есть несколько похожих(введите номер нужной статьи):\n';
    for (var i in res) {
        str +=(i*1+1)+') '+ res[i].name + '\n';
    }
    sendMess(id, str);
    reader(read1(id,callback,res),id)
}
function selection1(body, id,callback) {
    var $ = cheerio.load(body);
    var a = $('#mw-content-text').find('ul').find('li');
    var fst;
    var res = [];
   // console.log(a);
    for (var i in a) {
      //  console.log(i);
        if (!Number.isInteger(i * 1))continue;
        res.push({'name': '', 'url': ''});
        fst = a[i];
        var name=$(fst).find('a').html()||$(fst).find('a').attr('title');
        res[res.length - 1]['name'] =$(fst).find('a').attr('title');
        res[res.length - 1]['url'] = 'https://ru.wikipedia.org'+$(fst).find('a').attr('href');

    }
    var str = 'У данной фразы есть несколько значений(введите номер нужного):\n';
    for (var i in res) {
        str +=(i*1+1)+') '+ res[i].name + '\n';
    }
     //  console.log(str);
    sendMess(id, str);
    reader(read1(id,callback,res),id);
}

function think(body, id, callback,url) {

    var $ = cheerio.load(body);
    var a = $('.searchresults').find($('.mw-search-results')).find("li");
    var fH = getById(body, 'firstHeading');

    var head=$(fH).html();

    if (head.indexOf("(&#x437;&#x43D;&#x430;&#x447;&#x435;&#x43D;&#x438;&#x44F;)") + 1) {
       // console.log('here');
        //callback('(значения)');
        selection1(body,id,callback);
    } else if (a['0']) {
        selection(body, id,callback);
    } else if($('.searchresults').find($('.mw-search-nonefound')).html() ){
        callback('По запросу ничего не найдено');
    }
      else {
        callback(url);
    }
}

function find(name, id, callback, url) {

    if (!url) {
      //  console.log('нет урла');
        url = getUrl(name);
    }
    //console.log(url);
    request(url, function (error, response, body) {
        if (!error && response.statusCode == 200) {
           // console.log('find');

            think(body, id, callback,url);
        }
    });
}


exports.getR = function (r) {
    reader = r;
};
function read(id) {
    var res = {};
    sendMess(id, 'Что найти?');
    res.f = function (text) {
        var r = {};
        r.del = true;
        mainF(text, id);
        return r;
        //console.log(text);
    };
    return res;
}
var mainF = function (text, id) {
    if (text) {
        find(text, id, function (res) {
            sendMess(id, res)
        });
    } else {
        var res = read(id);
        reader(res, id);
    }
};
exports.mainF = mainF;
exports.info = function () {
    return 'Работает с wikipedia';
};
module.exports.addF = function () {
    return {};
};
///console.log(urlencode.decode('&#x437;&#x43D;&#x430;&#x447;&#x435;&#x43D;&#x438;&#x44F;'))
var sendPhoto;
exports.getPhoto=function(m){
    sendPhoto=m;
};
