// MyModules/SolveString/index.js
function leter(ch,pr,lN,rN,f){
    this.char=ch;
    this.pr=pr;
    this.lN=lN;
    this.rN=rN;
    this.f=f;
}
var operators=[];
operators.push(new leter('+',1,true,true,function(n1,n2){
    //   console.log(n1+' '+n2);
    return n1*1+n2*1;
}));
operators.push(new leter('-',0,true,true,function(n1,n2){
    return n1-n2;
}));
operators.push(new leter('*',5,true,true,function(n1,n2){
    return n1*n2;
}));
operators.push(new leter('/',5,true,true,function(n1,n2){
    return n1/n2;
}));
operators.push(new leter('^',10,true,true,function(n1,n2){
    return Math.pow(n1,n2);
}));
operators.push(new leter('cos',10,false,true,function(n1,n2){
    return Math.cos(n2);
}));
operators.push(new leter('sin',10,false,true,function(n1,n2){
    return Math.sin(n2);
}));
operators.push(new leter('tg',10,false,true,function(n1,n2){
    return Math.tan(n2);
}));
operators.push(new leter('ctg',10,false,true,function(n1,n2){
    return 1/(Math.tan(n2));
}));
function opInd(ch){
    for(var i in operators){
        if(operators[i].char===ch){
            return i;
        }
    }
    return -1;
}
function f(str){
    str=str.replace(/ /g,'');
    str=str.toLowerCase();
    //var t,p;
    var rBrack=[];
    var lBrack=[];
    for(var i=0;i<str.length;i++){
        if(str[i]==='('){
            rBrack.push(i);
        }
    }
    for(var i=0;i<str.length;i++){
        if(str[i]===')'){
            lBrack.push(i);
        }
    }
    if(rBrack.length!==lBrack.length){
        return 'Неправильный ввод';
    }
    for(var i= rBrack.length-1;i>-1;i--){
        var p;

        for(var j=rBrack[i];j<str.length;j++){
            //  console.log()
            if(str[j]==')'){
                p=j;
                break;
            }
        }
        var subS;
        //   console.log(rBrack[i]+' '+p);
        //if(rBrack[i]) {
        subS = str.substr(rBrack[i] + 1, p - rBrack[i] - 1);

        var c=calc(subS);
        //console.log(c);
        str=str.replace('('+subS+')',c);
        //    console.log(str);
    }
    return calc(str);
}
function calc(s){
    //while
    var syms=['+','/','*','-'];
    var sym=' ';
   while(sym){
        sym=null;
        for (var i = 0; i < operators.length; i++) {
            if(s.indexOf(operators[i].char)+1){
                //var ind=s.indexOf(operators[i].char);
                if(operators[i].char==='-' && s.indexOf(operators[i].char)===0){
                    if(s.slice(1).indexOf(operators[i].char)+1){

                    }else {
                        continue;
                    }
                }


                if(!sym ){
                    sym = i;
                }
                if( operators[i].pr > operators[sym].pr){
                    sym=i;
                }
            }

            /*if (opInd(s[i]) + 1) {
                if(!sym ){
                    sym = opInd(s[i]);
                }
                if( operators[opInd(s[i])].pr >= operators[sym].pr){
                    sym=opInd(s[i]);
                }
            }*/
        }
       // console.log(sym);
        if(sym===null) break;
        var n1 = '';

        var ind=s.indexOf(operators[sym].char);
        if(operators[sym].lN)
        n1=findLNum(s,ind-1);
        var n2 = '';
        //i = 1;
        if(operators[sym].rN)
        n2=findRNum(s,ind+operators[sym].char.length);

       // console.log(n1+'    '+n2);
        //n2=n2||'';
        //n1=n1||'';
        var res=operators[sym].f(n1*1,n2*1);
       //console.log(res);
       //console.log(n1.toString() + '' +operators[sym].char + ''+n2.toString());
        s = s.replace(n1 + '' +operators[sym].char + n2, res);
        //console.log(s);
    }
    return s;
}
function findLNum(str,num){
    var res;
    for(var i=num;i>=0;i--){
        if(!isNaN(str.slice(i,num+1)*1)){
            res=str.slice(i,num+1)*1;
        }
    }
    return res;
}
function findRNum(str,num){
    var res;
    for(var i=num;i<=str.length;i++){
        if(!isNaN(str.slice(num,i)*1)){
            res=str.slice(num,i)*1;
        }
    }
    return res;
}
//console.log(f('cos0'));
exports.solve=f;
