// ADS/TABLE/index.js
"use strict";
var name;
var deb=require('DEBUG').debug;

function debug(text){
    deb(text,name);
}
exports.getN=function(n){
    name=n;
};
var reader;
var sendMess;
var sendStick;
exports.getStick=function(s){
    sendStick=s;
};
exports.getMess=function(m){
    sendMess=m;
};
exports.getR=function(r){
    reader=r;
};
function read(id){
    var res={};
    sendMess(id, 'На какой день нужно расписание?');
    res.f=function(text){
        var r={};
        r.del=true;
        mainF(text,id);
        return r;
    };
    return res;
}
function mainF(text,id){
    function lesson(name){
        this.name=name;
    }
    if(!text){
        var res= read(id);
        reader(res,id);
        return;
    }
    text*=1;
    text--;
    if(text>=7){
        sendMess(id, 'В неделе всего 7 дней!');
    }

    var table=[];
    table.push([new lesson('математика'),new lesson('биология'),new lesson('физкультура'),new lesson('математика'),new lesson('математика'),new lesson('экономика')]);
    table.push([new lesson('обществознание'),new lesson('история'),new lesson('английский язык'),new lesson('физика'),new lesson('математика'),new lesson('математика'),new lesson('математика')]);
    table.push([new lesson('математика'),new lesson('математика'),new lesson('ОБЖ'),new lesson('физкультура'),new lesson('география'),new lesson('география')]);
    table.push([new lesson('английский язык'),new lesson('физика'),new lesson('физика'),new lesson('обществознание'),new lesson('химия'),new lesson('информатика')]);
    table.push([new lesson('литература'),new lesson('литература'),new lesson('математика'),new lesson('математика'),new lesson('физика'),new lesson('история')]);
    table.push([new lesson('русский язык'),new lesson('русский язык'),new lesson('физкультура'),new lesson('литература'),new lesson('английский язык'),new lesson('информатика'),new lesson('физика')]);
    var res='';
    if(!table[text]) {
        sendMess(id, 'На этот день нет расписания');
        return;
    }
    for(var i=0;i<table[text].length;i++){
        res+=(i+1)+') '+table[text][i].name;
        if(i!=table[text].length-1){
            res+='\n';
        }
    }
    sendMess(id,res);
};
exports.mainF=mainF;
exports.info=function(){
    return 'Расписание уроков';
};
exports.addF=function(){
    return {};
};
var sendPhoto;
exports.getPhoto=function(m){
    sendPhoto=m;
};
