// ADS/TABLE/index.js
"use strict";
var name;
var deb=require('DEBUG').debug;

function debug(text){
    deb(text,name);
}
exports.getN=function(n){
    name=n;
};
var reader;
var sendMess;
var sendStick;
exports.getStick=function(s){
    sendStick=s;
};
exports.getMess=function(m){
    sendMess=m;
};
exports.getR=function(r){
    reader=r;
};
function read(id){
    var res={};
    sendMess(id, 'На какой день нужно расписание?');
    res.f=function(text){
        var r={};
        r.del=true;
        mainF(text,id);
        return r;
    };
    return res;
}
function mainF(text,id){
    function lesson(name){
        this.name=name;
    }
    if(!text){
        var res= read(id);
        reader(res,id);
        return;
    }
    text*=1;
    text--;
    if(text>=7){
        sendMess(id, 'В неделе всего 7 дней!');
    }

    var table=[];
    table.push([new lesson('ОБЖ'),new lesson('Биология'),new lesson('Физкультура'),new lesson('Алгебра'),new lesson('Алгебра'),new lesson('экономика')]);
    table.push([new lesson('Химия'),new lesson('География'),new lesson('Алгебра'),new lesson('Алгебра'),new lesson('Информатика'),new lesson('История'),new lesson('Математика')]);
    table.push([new lesson('География'),new lesson('Англ. язык'),new lesson('Физика'),new lesson('Физкультура'),new lesson('Геометрия'),new lesson('Геометрия'),new lesson('Какоето говно по физике')]);
    table.push([new lesson('Обществознание'),new lesson('Обществознание'),new lesson('Литература'),new lesson('Литература'),new lesson('Физика'),new lesson('История')]);
    table.push([new lesson('Физика'),new lesson('Алгебра'),new lesson('Англ. язык'),new lesson('Математика'),new lesson('Литература'),new lesson('Геометрия')]);
    table.push([new lesson('Русский язык'),new lesson('Русский язык'),new lesson('Физкультура'),new lesson('Англ. язык'),new lesson('Физика'),new lesson('Информатика'),new lesson('Физика')]);
    var res='';
    if(!table[text]) {
        sendMess(id, 'На этот день нет расписания');
        return;
    }
    for(var i=0;i<table[text].length;i++){
        res+=(i+1)+') '+table[text][i].name;
        if(i!=table[text].length-1){
            res+='\n';
        }
    }
    sendMess(id,res);
};
exports.mainF=mainF;
exports.info=function(){
    return 'Расписание уроков';
};
exports.addF=function(){
    return {};
};
var sendPhoto;
exports.getPhoto=function(m){
    sendPhoto=m;
};
